dali/util/random_crop_generator.cc

```diff
@@ -23,66 +24,81 @@ RandomCropGenerator::RandomCropGenerator(
-
-        size_t original_area = H * W;
-        float target_area = scale * original_area;
-
-        int w, h;
-
-        // Uniform distribution is not suitable for aspect ratios. Here, we randomly
-        // draw either W/H or H/W aspec ratio which has a mean equal 1 for ranges with
-        // reciprocal aspect ratio ranges, such as 3/4..4/3
-        if (swap) {
-          float ratio = inv_aspect_ratio_dis_(rand_gen_);
-          w = static_cast<int>(
-              std::roundf(sqrtf(target_area / ratio)));
-          h = static_cast<int>(
-              std::roundf(sqrtf(target_area * ratio)));
-        } else {
-          float ratio = aspect_ratio_dis_(rand_gen_);
-          w = static_cast<int>(
-              std::roundf(sqrtf(target_area * ratio)));
-          h = static_cast<int>(
-              std::roundf(sqrtf(target_area / ratio)));
-        }
-
-        CropWindow crop;
-        if (w > 0 && h > 0 && w <= W && h <= H) {
-            float rand_x = uniform_(rand_gen_);
-            float rand_y = uniform_(rand_gen_);
-
-            crop.w = w;
-            crop.h = h;
-            crop.x = static_cast<int>(rand_x * (W - w));
-            crop.y = static_cast<int>(rand_y * (H - h));
-            return crop;
-        }
+  CropWindow crop = {};
+  if (W <= 0 || H <= 0) {
+    return crop;
+  }
+
+  float min_wh_ratio = aspect_ratio_range_.first;
+  float max_wh_ratio = aspect_ratio_range_.second;
+  float max_hw_ratio = 1 / aspect_ratio_range_.first;
+  float min_area = W * H * area_dis_.a();
+  int maxW = std::max<int>(1, H * max_wh_ratio);
+  int maxH = std::max<int>(1, W * max_hw_ratio);
+
+  // detect two impossible cases early
+  if (H * maxW < min_area) {  // image too wide
+    crop.h = H;
+    crop.w = maxW;
+  } else if (W * maxH < min_area) {  // image too tall
+    crop.w = W;
+    crop.h = maxH;
+  } else {
+    // it can still fail for very small images when size granularity matters
+    int attempts_left = num_attempts_;
+    for (; attempts_left > 0; attempts_left--) {
+      float scale = area_dis_(rand_gen_);
+
+      size_t original_area = H * W;
+      float target_area = scale * original_area;
+
+      float ratio = std::exp(aspect_ratio_log_dis_(rand_gen_));
+      crop.w = static_cast<int>(
+          std::roundf(sqrtf(target_area * ratio)));
+      crop.h = static_cast<int>(
+          std::roundf(sqrtf(target_area / ratio)));
+
+      if (crop.w < 1)
+        crop.w = 1;
+      if (crop.h < 1)
+        crop.h = 1;
+
+      ratio = static_cast<float>(crop.w) / crop.h;
+      if (crop.w <= W && crop.h <= H && ratio >= min_wh_ratio && ratio <= max_wh_ratio)
+        break;
     }
 
-    // If kMaxAttempts were consumed, use default crop
-    int min_dim = H < W ? H : W;
-    CropWindow crop;
-    crop.w = min_dim;
-    crop.h = min_dim;
-    crop.x = (W - min_dim) / 2;
-    crop.y = (H - min_dim) / 2;
-    return crop;
+    if (attempts_left <= 0) {
+      float max_area = area_dis_.param().b() * W * H;
```

Comment:
Why using `.param()` here and not at L59?

Reply:
`area_dis_.b()` getter should do it

Reply:
bug?

Reply:
Honestly, I didn't know there was a getter for it... Doesn't change much, though.

Reply:
Fixed.